### ISSUE TITLE
Custom Layer API is not working like functional API

### ISSUE THREAD

**Issue by somedadaism** [keras-team/keras#11653]
Please make sure that the boxes below are checked before you submit your issue.
If your issue is an **implementation question**, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [on the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) instead of opening a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
`pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps`

- [x] Check that your version of TensorFlow is up-to-date. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

I have implemented a custom Layer for a residual network but it failed to generalize, when I tested with simple MNIST.  
I thought the model was too simple to behave like this, so I additionally implemented the code again, using the functional API. This gives an improve from ~20% to 90% test accuracy.
Code:
```
class ResidualLayer(Layer):  # This does not generalize!

def __init__(self, conv_layer1, conv_layer2, **kwargs):
    self.__conv_layer1 = conv_layer1
    self.__conv_layer2 = conv_layer2
    super(ResidualLayer, self).__init__(**kwargs)

def build(self, input_shape):
    self.__in_shape = input_shape
    self.bn1 = BatchNormalization()
    self.bn2 = BatchNormalization()
    self.rel1 = Activation(activation='relu')
    self.rel2 = Activation(activation='relu')
    self.add = Add()
    self.scale_conv = Conv2D(filters=self.__conv_layer1.filters,
                             strides=self.__conv_layer1.strides,
                             kernel_size=(1, 1),
                             padding=self.__conv_layer1.padding)
    super(ResidualLayer, self).build(input_shape)

def call(self, x):
    input_layer = x
    tmp_layer = self.__conv_layer1(input_layer)
    tmp_layer = self.bn1(tmp_layer)
    tmp_layer = self.rel1(tmp_layer)
    tmp_layer = self.__conv_layer2(tmp_layer)
    tmp_layer = self.bn2(tmp_layer)
    if self.__conv_layer1.output_shape == self.__in_shape:
        tmp_layer = self.add([tmp_layer, input_layer])
    else:
        scaled_layer = self.scale_conv(input_layer)
        tmp_layer = self.add([tmp_layer, scaled_layer])
    tmp_layer = self.rel2(tmp_layer)
    return tmp_layer

def compute_output_shape(self, input_shape):
    return self.__conv_layer2.output_shape
```
```
def reslayer(x, conv_layer1, conv_layer2): # This works!
    input_layer = x
    tmp_layer = conv_layer1(input_layer)
    tmp_layer = BatchNormalization()(tmp_layer)
    tmp_layer = Activation(activation='relu')(tmp_layer)
    tmp_layer = conv_layer2(tmp_layer)
    tmp_layer = BatchNormalization()(tmp_layer)
    if conv_layer1.output_shape == keras.backend.shape(x):
        tmp_layer = Add()([tmp_layer, input_layer])
    else:
        scale_conv = Conv2D(filters=conv_layer1.filters, strides=conv_layer1.strides, kernel_size=(1, 1), padding=conv_layer1.padding)
        scaled_layer = scale_conv(input_layer)
        tmp_layer = Add()([tmp_layer, scaled_layer])
        tmp_layer = Activation(activation='relu')(tmp_layer)
    return tmp_layer
```

I would like to understand where the mistake with using the Custom Layer API happened.

**Comment by omalleyt12** [comment]
Keras Layers currently don't support having sublayers inside of them. You can subclass the `Model` class and use `ResidualLayer` as a nested submodel

**Comment by somedadaism** [comment]
I think this is dangerous for the user, because it is an easy way to use the Sequential API, is there a way to give a warning? Mind that this code was executing, but not running properly for the validation set.

**Comment by omalleyt12** [comment]
Agreed. We're working on a new approach for this, to either allow layers to have sublayers, or to throw a warning when this happens. I'll update this thread when the solution's in, please feel free to comment ideas

**Comment by IwraStudios** [comment]
I'm have also run into this problem. I'm working on a custom layer, which uses 3 Conv2D models and transforms them using `keras.backend` and `keras.tf` functions combined with some trained weights.
At first I thought I could maybe include the weights manually using `self.add_weight`, but that also doesn't seem to be possible. It appears to remove them from the graph completely. I'm looking forward to the solution (or alternatively an explicit "unsupported" statement)
